nsv13/code/controllers/subsystem/starsystem.dm

```diff
@@ -86,17 +91,130 @@ Returns a faction datum by its name (case insensitive!)
+	}
+
+	catch(var/exception/ex){
+		//Fallback: Load the hardcoded systems and report an error.
+		instantiate_systems_backup()
+		CRASH("FATAL: Unable to load starmap from: [_source_path]. (Defaulting...): [ex]")
+	}
+
+	for(var/i = 1; i <= _systems.len; i++)
+		//Try instancing this system from JSON, jump out if anything goes wrong.
+		var/list/sys_info = _systems[i]
+		try{
+			//Fields that must be present are accessed unsafely. The rest is fine if null.
+			var/datum/star_system/next = new /datum/star_system(
+				//Required props, you must fill these out, or it's an invalid map.
+				name = sys_info["name"],
+				desc = sys_info["desc"],
+				x = sys_info["x"],
+				y = sys_info["y"],
+				alignment = sys_info["alignment"],
+				hidden = sys_info["hidden"],
+				sector = sys_info["sector"],
+				adjacency_list = json_decode(sys_info["adjacency_list"]),
+				//Optional props. Recommended, but can be left blank.
+				threat_level = LAZYACCESS(sys_info, "threat_level") || THREAT_LEVEL_NONE,
+				is_capital = LAZYACCESS(sys_info, "is_capital") || FALSE,
+				fleet_type = LAZYACCESS(sys_info, "fleet_type") || null,
+				parallax_property = LAZYACCESS(sys_info, "parallax_property") || null,
+				visitable = LAZYACCESS(sys_info, "visitable") || TRUE,
+				is_hypergate = LAZYACCESS(sys_info,"is_hypergate") || FALSE,
+				preset_trader = (LAZYACCESS(sys_info,"preset_trader")) ? text2path(sys_info["preset_trader"]) : null,
+				system_traits = LAZYACCESS(sys_info,"system_traits") || list(),
+				system_type = (LAZYACCESS(sys_info,"system_type") && sys_info["system_type"] != "null" && sys_info["system_type"] != null) ? json_decode(sys_info["system_type"]) : list(),
+				audio_cues = (LAZYACCESS(sys_info,"audio_cues") && sys_info["audio_cues"] != "null" && sys_info["audio_cues"] != null) ? json_decode(sys_info["audio_cues"]) : list(),
+				wormhole_connections = (LAZYACCESS(sys_info,"wormhole_connections") && sys_info["wormhole_connections"] != "null" && sys_info["wormhole_connections"] != null) ? json_decode(sys_info["wormhole_connections"]) : list(),
+				startup_proc = LAZYACCESS(sys_info,"startup_proc") || null
+				//Future implementation ideas: We can cache system_contents. I'm not dealing with that now though.
+			)
+			systems += next
+		}
+		catch(var/exception/e){
+			message_admins("WARNING: Invalid star system in json: [sys_info["name"]] ([e]). Skipping...")
+			continue
+		}
+	message_admins("Successfully loaded starmap layout from [_source_path]")
+
+
+/datum/controller/subsystem/star_system/proc/instantiate_systems_backup()
 	for(var/instance in subtypesof(/datum/star_system))
 		var/datum/star_system/S = new instance
 		if(S.name)
 			systems += S
+	if(saving)
+		return
+	saving = TRUE
+	save()
+	saving = FALSE
+/**
+<summary>Save the current starmap layout to a json file. Used for persistence.</summary>
+<param></param>
+*/
+
+/datum/controller/subsystem/star_system/proc/save(_destination_path = "config/starmap/starmap_default.json")
```

Comment:
this is very dangerous, the file path, type or proc caller aren't checked at all. And throw fdel into the mix... 💀

Reply:
Yeah god that's horrible you're right 😂